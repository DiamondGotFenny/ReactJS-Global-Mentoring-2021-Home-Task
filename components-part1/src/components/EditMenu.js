import React, { useState, memo } from 'react';
import { useDispatch } from 'react-redux';
import {useMutation } from 'react-query';
import styled from 'styled-components';
import MovieModal from './MovieModal';
import httpService from '../services/httpService';
import { deleteMovie } from '../actions/moviesActions';

const StyledMenu = styled.nav`
  display: ${({ open }) => (open ? 'inline-flex;' : 'none;')};
  flex-direction: column;
  justify-content: center;
  color: #fff;
  background: #effffa;
  transform: translate(4rem, -13rem);
  width: 190px;
  height: 111px;
  background: rgba(35, 35, 35, 0.918051);
  box-shadow: 0px 0px 2px rgba(0, 0, 0, 0.1), 0px 2px 10px rgba(0, 0, 0, 0.1),
    0px 10px 20px rgba(0, 0, 0, 0.1), 0px 10px 50px rgba(0, 0, 0, 0.1);
  backdrop-filter: blur(27.1828px);
  border-radius: 4px;
  position: absolute;

  & button {
    color: #fff;
    height: 50%;
    background: transparent;
    border: none;
    &: hover {
      background: #f65261;
    }
  }
  & .close {
    display: inline-flex;
    position: absolute;
    top: 1rem;
    right: 1rem;
    background: transparent;
  }
  z-index: 10;
`;
const deleteMovieReq= async (id) => {
  const {data} = await httpService.delete(`/movies/${id}`);
  return data;
}
const EditMenu = ({ open, setOpen, movie }) => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const dispatch = useDispatch();

  const {isLoading:isDeletingMovie,mutate: deleteMovieReq} = useMutation(
    async () => {
      const {data} = await httpService.delete(`/movies/${movie.id}`);
      return data;
    },
   { enabled:false,
    onSuccess: () => {
      dispatch(deleteMovie(movie.id));
    },
    onError: (error) => {
      alert(error.message);
    },
  }

  );

  const handleClose = () => {
    setIsModalOpen(false);
  };
  const handleOpen = () => {
    setIsModalOpen(true);
  };
/*   const handleDelete = async (id) => {
    try {
      await httpService.delete(`/movies/${id}`);
      dispatch(deleteMovie(id));
    } catch (error) {
      alert(error);
    }
  }; */
  const renderForm = React.useMemo(() => {
    return (
      <MovieModal
        isOpen={isModalOpen}
        handleClose={handleClose}
        movie={movie}
      />
    );
  }, [movie, isModalOpen, dispatch]);
  return (
    <>
      <StyledMenu open={open}>
        <button className="close" onClick={() => setOpen(false)}>
          X
        </button>
        <button onClick={handleOpen}>Edit</button>
        <button onClick={() => deleteMovieReq()}>Delete</button>
      </StyledMenu>
      {renderForm}
    </>
  );
};

export default memo(EditMenu);
