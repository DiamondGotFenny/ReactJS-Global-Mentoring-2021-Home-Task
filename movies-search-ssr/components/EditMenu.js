import React, { memo, useState } from 'react';

import httpService from '../services/httpService';
import styles from './EditMenu.module.css';

const EditMenu = ({ open, setOpen, movie }) => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const handleClose = () => {
    setIsModalOpen(false);
  };
  const handleOpen = () => {
    setIsModalOpen(true);
  };
  const handleDelete = async (id) => {
    try {
      await httpService.delete(`/movies/${id}`);
    } catch (error) {
      alert(error);
    }
  };

  return (
    <>
      <nav className={open ? styles.Editmenu : styles.hidden}>
        <button className="close" onClick={() => setOpen(false)}>
          X
        </button>
        <button onClick={handleOpen}>Edit</button>
        <button onClick={() => handleDelete(movie.id)}>Delete</button>
      </nav>
    </>
  );
};

export default EditMenu;
