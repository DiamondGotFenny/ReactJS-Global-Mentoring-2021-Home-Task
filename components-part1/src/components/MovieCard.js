import React, { memo } from 'react';
import styled from 'styled-components';
import EditMenu from './EditMenu';
import { useDispatch } from 'react-redux';
import { fetchMovieDetails } from '../actions/moviesActions';
const StyledCardContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 323px;
  font-family: Montserrat;
  font-style: normal;
  font-weight: 500;
  font-size: 18px;
  line-height: 22px;
  /* identical to box height */

  color: #ffffff;
  cursor: pointer;
  mix-blend-mode: normal;
  opacity: 0.7;
  margin: 26px;

  img {
    width: 323px;
  }

  &:hover .menuButton {
    display: flex;
  }
`;

const BasicInfoWrapper = styled.div`
  margin-top: 26px;
  display: flex;
  width: 100%;
  justify-content: space-between;
  & title {
    display: flex;
  }
`;

const GenreWrapper = styled.div`
  display: flex;
  width: 100%;
  justify-content: start;
  margin-top: 6px;
`;
const StyledDotsButton = styled.button`
  position: absolute;
  transform: translate(8rem, -15rem);
  display: none;
  width: 2rem;
  height: 2rem;
  flex-direction: column;
  justify-content: space-around;
  background: #2a202d;
  border-radius: 50%;
  border: none;
  cursor: pointer;
  padding: 0.3rem;
  z-index: 5;

  &:focus {
    outline: none;
  }

  div {
    width: 0.25rem;
    height: 0.25rem;
    background: #fff;
    border-radius: 10px;
    transition: all 0.3s linear;
    position: relative;
    transform-origin: 1px;
    left: 0.6rem;
  }
`;
const MovieCard = ({ movie }) => {
  const { title, poster_path, genres, release_date } = movie;
  const [open, setOpen] = React.useState(false);
  const dispatch = useDispatch();

  return (
    <StyledCardContainer>
      <StyledDotsButton
        className="menuButton"
        open={!open}
        onClick={() => setOpen(true)}
      >
        <div />
        <div />
        <div />
      </StyledDotsButton>
      <EditMenu open={open} setOpen={setOpen} movie={movie} />
      <div onClick={() => dispatch(fetchMovieDetails(movie))}>
        <img src={poster_path} alt={title} />
        <BasicInfoWrapper>
          <title>{title}</title>
          <span>{release_date}</span>
        </BasicInfoWrapper>
        <GenreWrapper>{genres.join(',')}</GenreWrapper>
      </div>
    </StyledCardContainer>
  );
};
export default memo(MovieCard);
