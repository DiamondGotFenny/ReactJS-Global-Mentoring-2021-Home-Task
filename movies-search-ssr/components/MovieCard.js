import React, { memo } from 'react';
import EditMenu from './EditMenu';
import router from 'next/router';
import styles from './MovieCard.module.css';

const MovieCard = ({ movie }) => {
  const { title, poster_path, genres, release_date } = movie;
  const [open, setOpen] = React.useState(false);

  return (
    <div className={styles['movie-card']}>
      <button
        className={styles.menuButton}
        open={!open}
        onClick={() => setOpen(true)}
      >
        <div />
        <div />
        <div />
      </button>
      <EditMenu open={open} setOpen={setOpen} movie={movie} />
      <div
        className="click-container"
        onClick={() => router.push(`/search?movie=${movie.id}`)}
      >
        <img src={poster_path} alt={title} />
        <div className={styles.BasicInfoWrapper}>
          <title>{title}</title>
          <span>{release_date}</span>
        </div>
        <div className={styles.GenreWrapper}>{genres.join(',')}</div>
      </div>
    </div>
  );
};
export default memo(MovieCard);
