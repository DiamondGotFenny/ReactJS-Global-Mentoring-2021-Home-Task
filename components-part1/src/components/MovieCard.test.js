import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import EditMenu from './EditMenu';
import MovieCard from './MovieCard';

const mockMovie = {
  id: 284053,
  title: 'Thor: Ragnarok',
  tagline: 'No Hammer. No Problem.',
  vote_average: 7.4,
  vote_count: 5349,
  release_date: '2017-10-25',
  poster_path:
    'https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg',
  overview:
    'Thor is imprisoned on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.',
  budget: 180000000,
  revenue: 854229371,
  genres: ['Action', 'Adventure', 'Fantasy'],
  runtime: 130,
};

const mockedUsedNavigate = jest.fn();
const mockUsedLocation = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
  useLocation: () => mockUsedLocation,
}));

const setup = (props = {}) => {
  const wrapper = shallow(<MovieCard {...props} />);
  return wrapper;
};

describe('render movie card correctly', () => {
  const wrapper = setup({ movie: mockMovie });
  it('should render without crash', () => {
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render movie info correctly', () => {
    expect(wrapper.find({ children: mockMovie.title })).toBeTruthy();
    expect(wrapper.find({ children: mockMovie.genres })).toBeTruthy();
    expect(wrapper.find({ children: mockMovie.release_date })).toBeTruthy();
    const imgSrc = wrapper.find('img').prop('src');
    expect(imgSrc).toBe(mockMovie.poster_path);
  });
});

describe('the movie card interactive correctly', () => {
  const wrapper = setup({ movie: mockMovie });
  it('should call the navigate function when clicked', () => {
    wrapper.find('.click-container').simulate('click');
    expect(mockedUsedNavigate).toHaveBeenCalledWith(
      `/search?movie=${mockMovie.id}`
    );
  });
  it('the EditMenu should be rendered when menu Button is clicked', () => {
    wrapper.find('.menuButton').simulate('click');
    expect(wrapper.containsMatchingElement(<EditMenu />)).toBeTruthy();
  });
});
