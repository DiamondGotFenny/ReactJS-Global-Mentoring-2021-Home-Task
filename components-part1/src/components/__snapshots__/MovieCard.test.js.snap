// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render movie card correctly should render without crash 1`] = `
<styled.div
  className="movie-card"
>
  <styled.button
    className="menuButton"
    onClick={[Function]}
    open={true}
  >
    <div />
    <div />
    <div />
  </styled.button>
  <Memo(EditMenu)
    movie={
      Object {
        "budget": 180000000,
        "genres": Array [
          "Action",
          "Adventure",
          "Fantasy",
        ],
        "id": 284053,
        "overview": "Thor is imprisoned on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
        "poster_path": "https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
        "release_date": "2017-10-25",
        "revenue": 854229371,
        "runtime": 130,
        "tagline": "No Hammer. No Problem.",
        "title": "Thor: Ragnarok",
        "vote_average": 7.4,
        "vote_count": 5349,
      }
    }
    open={false}
    setOpen={[Function]}
  />
  <div
    className="click-container"
    onClick={[Function]}
  >
    <img
      alt="Thor: Ragnarok"
      src="https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg"
    />
    <styled.div>
      <title>
        Thor: Ragnarok
      </title>
      <span>
        2017-10-25
      </span>
    </styled.div>
    <styled.div>
      Action,Adventure,Fantasy
    </styled.div>
  </div>
</styled.div>
`;
